e-ng-change not working for bsdate element

Hi,

Please check out my plunker -> http://plnkr.co/edit/KzAwC1W0S7SZaG0wghl0

I've got a table form which contains a datepicker column. I want to utilize the e-ng-change method to do some work every time the date changes which works for basic text elements but not bsdate elements. I keep getting the below error when going to edit the table form:

> VM1830 angular.min.js:116 Error: [$compile:ctreq] http://errors.angularjs.org/1.5.3/$compile/ctreq?p0=ngModel&p1=ngChange
>     at Error (native)
>     at https://ajax.googleapis.com/ajax/libs/angularjs/1.5.3/angular.min.js:6:416
>     at jb (https://ajax.googleapis.com/ajax/libs/angularjs/1.5.3/angular.min.js:70:406)
>     at n (https://ajax.googleapis.com/ajax/libs/angularjs/1.5.3/angular.min.js:65:259)
>     at g (https://ajax.googleapis.com/ajax/libs/angularjs/1.5.3/angular.min.js:58:140)
>     at g (https://ajax.googleapis.com/ajax/libs/angularjs/1.5.3/angular.min.js:58:157)
>     at g (https://ajax.googleapis.com/ajax/libs/angularjs/1.5.3/angular.min.js:58:157)
>     at https://ajax.googleapis.com/ajax/libs/angularjs/1.5.3/angular.min.js:57:281
>     at EditableController.self.show (http://vitalets.github.io/angular-xeditable/dist/js/xeditable.js:716:30)
>     at http://vitalets.github.io/angular-xeditable/dist/js/xeditable.js:1174:26

I've tried adding e-ng-model but still to no avail. Any guidance would be greatly appreciated.

--- a/docs/demos/bsdate/desc.md
+++ b/docs/demos/bsdate/desc.md
@@ -8,4 +8,6 @@ Add `ui.bootstrap` as module dependency:
 	var app = angular.module("app", ["xeditable", "ui.bootstrap"]);
 
 And set `editable-bsdate` attribute in editable element.
+To make the input field read-only and force the date to be selected from the popup, add the `e-readonly="true"` attribute.
+Add `e-ng-change` attribute to call a function when the value of the datepicker is changed.
 Other parameters can be defined via `e-*` syntax, e.g. `e-datepicker-popup="dd-MMMM-yyyy"`.

--- a/docs/demos/dev-bsdate/controller.js
+++ b/docs/demos/dev-bsdate/controller.js
@@ -25,4 +25,9 @@ app.controller('DevBsdateCtrl', function($scope) {
 
 		$scope.opened[elementOpened] = !$scope.opened[elementOpened];
 	};
+	
+	$scope.changed = function() {
+		$scope.user.changed = true;	
+		window.console.log("value changed");
+	};
 });

--- a/docs/demos/dev-bsdate/test.js
+++ b/docs/demos/dev-bsdate/test.js
@@ -113,4 +113,34 @@ describe('bsdate', function() {
     expect(element(s+'a#readOnly').text()).toMatch('02/05/2006');
     expect(element(s+'form').count()).toBe(0);
   });
+
+  it('should show editor and submit new value and on change event is called', function() {
+    var s = '[ng-controller="DevBsdateCtrl"] ';
+
+    expect(element(s+'a#onChangeDate').css('display')).not().toBe('none');
+    expect(element(s+'a#onChangeDate').text()).toMatch('15/05/1984');
+    element(s+'a#onChangeDate').click();
+    element(s+'form .input-group-btn button[type="button"]').click();
+
+    expect(element(s+'a#onChangeDate').css('display')).toBe('none');
+    expect(element(s+'form[editable-form="$form"]').count()).toBe(1);
+    expect(element(s+'form input[type="text"]:visible').count()).toBe(1);
+    expect(element(s+'form input[type="text"]').val()).toBe('15-May-1984');
+    expect(element(s+'form .editable-buttons button[type="submit"]:visible').count()).toBe(1);
+    expect(element(s+'form .editable-buttons button[type="button"]:visible').count()).toBe(1);
+    expect(element(s+'ul.dropdown-menu:visible').count()).toBe(1);
+    expect(element(s+'form table button.btn-info span').text()).toMatch('15');
+
+    //set 29 april
+    element(s+'form table > tbody > tr:eq(0) > td:eq(1) > button').click();
+    expect(element(s+'ul.dropdown-menu:visible').count()).toBe(0);
+    expect(element(s+'form input[type="text"]').val()).toBe('29-April-1984');
+
+    //submit
+    element(s+'form button[type="submit"]').click();
+
+    expect(element(s+'a#onChangeDate').css('display')).not().toBe('none');
+    expect(element(s+'a#onChangeDate').text()).toMatch('29/04/1984');
+    expect(element(s+'form').count()).toBe(0);
+  });
 });

--- a/docs/demos/dev-bsdate/view.html
+++ b/docs/demos/dev-bsdate/view.html
@@ -1,35 +1,45 @@
 <div ng-controller="DevBsdateCtrl">
-  <a href="#" editable-bsdate="user.dob"
-              e-is-open="opened.$data"
-              e-ng-click="open($event,'$data')"
-              e-datepicker-popup="dd-MMMM-yyyy"
-              e-min-date="getMinDate()"
-              e-max-date="getMaxDate()"
-              e-on-open-focus="true"
+  <a href="#" data-editable-bsdate="user.dob"
+              data-e-is-open="opened.$data"
+              data-e-ng-click="open($event,'$data')"
+              data-e-datepicker-popup="dd-MMMM-yyyy"
+              data-e-min-date="getMinDate()"
+              data-e-max-date="getMaxDate()"
+              data-e-on-open-focus="true"
               id="minMax"
               class="minMax">
     {{ (user.dob | date:"dd/MM/yyyy") || 'empty' }}
   </a>
-
-  <a href="#" editable-bsdate="user.startDate"
-              e-is-open="opened.$data"
-              e-ng-click="open($event,'$data')"
-              e-datepicker-popup="dd-MMMM-yyyy"
-              e-init-date="getInitDate()"
+  &nbsp;
+  <a href="#" data-editable-bsdate="user.startDate"
+              data-e-is-open="opened.$data"
+              data-e-ng-click="open($event,'$data')"
+              data-e-datepicker-popup="dd-MMMM-yyyy"
+              data-e-init-date="getInitDate()"
               id="initDate"
               class="initDate">
     {{ (user.startDate | date:"dd/MM/yyyy") || 'empty' }}
   </a>
-
-  <a href="#" editable-bsdate="user.hireDate"
-              e-is-open="opened.$data"
-              e-ng-click="open($event,'$data')"
-              e-datepicker-popup="dd-MMMM-yyyy"
-              e-init-date="getInitDate()"
-              e-on-open-focus="true"
-              e-readonly="true"
+  &nbsp;
+  <a href="#" data-editable-bsdate="user.hireDate"
+              data-e-is-open="opened.$data"
+              data-e-ng-click="open($event,'$data')"
+              data-e-datepicker-popup="dd-MMMM-yyyy"
+              data-e-init-date="getInitDate()"
+              data-e-on-open-focus="true"
+              data-e-readonly="true"
               id="readOnly"
               class="readOnly">
     {{ (user.hireDate | date:"dd/MM/yyyy") || 'empty' }}
+  </a>
+  &nbsp;
+  <a href="#" data-editable-bsdate="user.dob"
+              data-e-is-open="opened.$data"
+              data-e-ng-click="open($event,'$data')"
+              data-e-datepicker-popup="dd-MMMM-yyyy"
+              data-e-ng-change="changed()"
+              id="onChangeDate"
+              class="onChangeDate">
+    {{ (user.dob | date:"dd/MM/yyyy") || 'empty' }}
   </a>
 </div>

--- a/docs/demos/dev-select/controller.js
+++ b/docs/demos/dev-select/controller.js
@@ -1,6 +1,7 @@
 app.controller('DevSelectCtrl', function($scope, $filter) {
   $scope.user = {
-    status: [2, 4]
+    status: [2, 4],
+    status2: null
   };
 
   $scope.statuses = [
@@ -18,5 +19,10 @@ app.controller('DevSelectCtrl', function($scope, $filter) {
       }
     });
     return selected.length ? selected.join(', ') : 'Not set';
+  };
+  
+  $scope.showStatus2 = function() {
+    var selected = $filter('filter')($scope.statuses, {value: $scope.user.status2});
+    return ($scope.user.status2 && selected.length) ? selected[0].text : 'Not set';
   };
 });

--- a/docs/demos/dev-select/test.js
+++ b/docs/demos/dev-select/test.js
@@ -7,11 +7,11 @@ describe('dev-select-multiple', function() {
   it('should show options defined without value', function() {
     var s = '[ng-controller="DevSelectCtrl"] ';
 
-    expect(element(s+'a').text()).toMatch('status2');
-    expect(element(s+'a').text()).toMatch('status4');
-    element(s+'a').click();
+    expect(element(s+'a#multiSelect').text()).toMatch('status2');
+    expect(element(s+'a#multiSelect').text()).toMatch('status4');
+    element(s+'a#multiSelect').click();
 
-    expect(element(s+'a').css('display')).toBe('none');
+    expect(element(s+'a#multiSelect').css('display')).toBe('none');
     expect(element(s+'form[editable-form="$form"]').count()).toBe(1);
     expect(element(s+'form select:visible:enabled').count()).toBe(1);
     expect(element(s+'form select option').count()).toBe(4);
@@ -21,10 +21,30 @@ describe('dev-select-multiple', function() {
     using(s).select('$data').option('status2');
     element(s+'form button[type="submit"]').click();
 
-    expect(element(s+'a').css('display')).not().toBe('none');
-    expect(element(s+'a').text()).toMatch('status2');
-    expect(element(s+'a').text()).not().toMatch('status4');
+    expect(element(s+'a#multiSelect').css('display')).not().toBe('none');
+    expect(element(s+'a#multiSelect').text()).toMatch('status2');
+    expect(element(s+'a#multiSelect').text()).not().toMatch('status4');
     expect(element(s+'form').count()).toBe(0);
   });
 
+  it('should show options with default value selected', function() {
+    var s = '[ng-controller="DevSelectCtrl"] ';
+
+    expect(element(s+'a#defaultValue').text()).toMatch('Not set');
+    element(s+'a#defaultValue').click();
+
+    expect(element(s+'a#defaultValue').css('display')).toBe('none');
+    expect(element(s+'form[editable-form="$form"]').count()).toBe(1);
+    expect(element(s+'form select:visible:enabled').count()).toBe(1);
+    expect(element(s+'form select option').count()).toBe(5);
+    expect(element(s+'form select option:selected').count()).toBe(1);
+    expect(element(s+'form select').val()).toMatch('');
+
+    using(s).select('$data').option('status2');
+    element(s+'form button[type="submit"]').click();
+
+    expect(element(s+'a#defaultValue').css('display')).not().toBe('none');
+    expect(element(s+'a#defaultValue').text()).toMatch('status2');
+    expect(element(s+'form').count()).toBe(0);
+  });
 });

--- a/docs/demos/dev-select/view.html
+++ b/docs/demos/dev-select/view.html
@@ -1,5 +1,15 @@
 <div ng-controller="DevSelectCtrl">
-  <a href="#" editable-select="user.status" e-multiple e-ng-options="s.value as s.text for s in statuses">
+  <a href="#" data-editable-select="user.status"
+              data-e-multiple
+              data-e-ng-options="s.value as s.text for s in statuses"
+              id="multiSelect"
+              name="multiSelect">
     {{ showStatus() }}
   </a>
+  &nbsp;
+  <a href="#" data-editable-select="user.status2"
+              data-e-ng-options="s.value as s.text for s in statuses"
+              data-e-placeholder="- Select Status -"
+              id="defaultValue"
+              name="defaultValue">{{showStatus2()}}</a>
 </div>  

--- a/docs/demos/select-local/desc.md
+++ b/docs/demos/select-local/desc.md
@@ -1,3 +1,4 @@
 To create editable select (dropdown) just set `editable-select` attribute pointing to model.
 To pass dropdown options you should define `e-ng-options` attribute
 that works like normal angular `ng-options` but is transfered to underlying `<select>` from original element.
+To set a default option in the list add the `e-placeholder` attribute.

--- a/src/js/directives/bsdate.js
+++ b/src/js/directives/bsdate.js
@@ -14,7 +14,7 @@ angular.module('xeditable').directive('editableBsdate', ['editableDirectiveFacto
                  **/
                 this.parent.render.call(this);
 
-                var inputDatePicker = angular.element('<input type="text" class="form-control" ng-model="$data"/>');
+                var inputDatePicker = angular.element('<input type="text" class="form-control" data-ng-model="$data"/>');
                 var buttonDatePicker = angular.element('<button type="button" class="btn btn-default"><i class="glyphicon glyphicon-calendar"></i></button>');
                 var buttonWrapper = angular.element('<span class="input-group-btn"></span>');
 
@@ -33,6 +33,11 @@ angular.module('xeditable').directive('editableBsdate', ['editableDirectiveFacto
                 inputDatePicker.attr('name', this.attrs.eName);
                 inputDatePicker.attr('on-open-focus', this.attrs.eOnOpenFocus || true);
                 inputDatePicker.attr('ng-readonly', this.attrs.eReadonly || false);
+
+                if (this.attrs.eNgChange) {
+                    inputDatePicker.attr('ng-change', this.attrs.eNgChange);
+                    this.inputEl.removeAttr('ng-change');
+                }
 
                 this.scope.dateOptions = {
                     formatDay:  this.attrs.eFormatDay || 'dd',

--- a/src/js/directives/select.js
+++ b/src/js/directives/select.js
@@ -4,6 +4,14 @@ angular.module('xeditable').directive('editableSelect', ['editableDirectiveFacto
     return editableDirectiveFactory({
       directiveName: 'editableSelect',
       inputTpl: '<select></select>',
+      render: function() {
+        this.parent.render.call(this);
+
+        if (this.attrs.ePlaceholder) {
+          var placeholder = angular.element('<option value="">' + this.attrs.ePlaceholder + '</option>');
+          this.inputEl.append(placeholder);
+        }
+      },
       autosubmit: function() {
         var self = this;
         self.inputEl.bind('change', function() {

